### DISCUSSION TITLE
Empty file emitted for `await import('crypto')` in a dependency despite resolve.fallback.crypto = false

### DISCUSSION THREAD

**Discussion by knightcode** [webpack/webpack#18028] [Q&A]
# Bug report

**What is the current behavior?**

After upgrading a package---in this case, `bson` from 4.x to 5.x/6.x---webpack began emitting an additional chunk, named `_5ab2.js` in my case. And the URL used to try and fetch this file from the server doesn't seem to properly respect my `output.publicPath` config. The `publicPath` prefix is prepended twice.

**If the current behavior is a bug, please provide the steps to reproduce.**

After finding the reference to this file in the main entry point's generated code, I traced it back to a function in the bson package:

```javascript
const nodejsRandomBytes = await (async () => {
    try {
        return (await import('crypto')).randomBytes;
    }
    catch {
        return nodejsMathRandomBytes;
    }
})();
```

My webpack.config.js has had `resolve: { fallback: { crypto: false } }` for extensions `[".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs"]`, so that I don't believe it should be trying to import anything. Moreover, the emitted file is essentially empty:

```javascript
(self["webpackChunkscserver"] = self["webpackChunkscserver"] || []).push([["_5ab2"],{

/***/ "?5ab2":
/*!************************!*\
  !*** crypto (ignored) ***!
  \************************/
/***/ (() => {

/* (ignored) */

/***/ })

}]);
//# sourceMappingURL=_5ab2.js.map
```

When trying to load this file, the server gets a request for `/js/js/_5ab2.js`, to which it responds with a 404.

Relevant Config:
```javascript
  const frontendConfig = {
    experiments: { topLevelAwait: true },
    devtool: "source-map",
    resolve: {
      extensions: [".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs"],
      fallback: {
        crypto: false
      }
    },
    entry: {
      app: "./src/frontend/ts/main.tsx"
    },
    output: {
      path: path.join(__dirname, "public/js"),
      publicPath: "js/",
      filename: "[name].js",
      chunkFilename: "[name].js"
    },
    module: {
      rules: [{
        test: /src\/.*svg$/,
        use: [ "raw-loader" ]
      },
      ...
      ,{
        test: /\.tsx?$/,
        exclude: /node_modules/,
        use: [{
          loader: "ts-loader",
          options: {configFile: path.resolve(__dirname, "src", "frontend", "tsconfig.json")}
        }]
      }, {
        test: /\.wasm$/,
        type: "asset/resource"
      }]
    },
    externals: {
      "react": "React",
      "react-dom": "ReactDOM"
    },
    plugins: [
      new webpack.DefinePlugin({
        secrets: ...
      }),
      new MiniCssExtractPlugin({
        filename: "extracted_styles.css"
      })
    ]
  };
```

my package.json has `"type": "module"`.

I have tried `require.resolve('crypto-browserify')` in the config, too, but the URL is still erroneous (and the output file is just a lot more comments... due to tree shaking, I think).

**What is the expected behavior?**

Ideally, no extra file would be emitted since it's unnecessary. MongoDB's documentation for `bson` and the presence of the try/catch block all suggest an intention that the import should fail in the browser. 

If the empty file must be emitted, though, the URL should be correct when the system tries to fetch it.

**Other relevant information:**
webpack version:  5.90.0
Node.js version: 18.19.0
Operating System: macOS 14.2.1
Additional tools: webpack-cli@5.1.4

**Comment by alexander-akait** [comment]
Just remove `publicPath: "js/"`, webpack 5 supports automatically public path resolution (i.e. use relative paths)

**Reply by alexander-akait** [reply]
If it didn't help, please create a small reproducible test repo

**Reply by knightcode** [reply]
so the clutter is expected? the extra fetch is acceptable?

**Reply by alexander-akait** [reply]
@knightcode If you want to merge chunks, you can use https://webpack.js.org/plugins/split-chunks-plugin/ and merge it into the one initial chunk, yeah, extra fetch is expected, because each dynamic `import` is a chunk.

If you need help with `SplitChunksPlugin`, feel free to ask

**Reply by knightcode** [reply]
@alexander-akait  that makes sense, but two things:

1. the presence of `resolve: { fallback: { crypto: false } }` in the config in this case. Am I not telling webpack to ignore imports of `crypto` here?
2. if it tree-shakes down to nothing (I assume that's what's happening), ...optimize

**Reply by alexander-akait** [reply]
> the presence of resolve: { fallback: { crypto: false } } in the config in this case. Am I not telling webpack to ignore imports of crypto here?

Yeah, you just say - this module will be empty

> if it tree-shakes down to nothing (I assume that's what's happening), ...optimize

It depends on some conditions, but you can always specify - merge this module. 

Perhaps the developers should improve this code and check the node environment before creating a request.
